README.md
## Sift PHP framework

                o88     o888o  o8
      oooooooo8 oooo  o888oo o888oo
     888ooooooo  888   888    888
             888 888   888    888
     88oooooo88 o888o o888o    888o


## Requirements

 * PHP > 5.2.3


CHANGELOG
08.11.2013 [9d158e6] added slovak translation for 500 and unavailable error page
06.11.2013 [5dc6346] allow HEAD request in view cache manager
06.11.2013 [18ecee7] updated the skeleton htaccess - allow caching of resources
06.11.2013 [acfafc1] updated minifier to allow client caching
05.11.2013 [425b3a8] added dummy js files which triggers an error in assetPackager filter
04.11.2013 [3b5b4e5] exception setAccessible() is available only on php > 5.3
04.11.2013 [3ee882c] fixed compat issues under php 5.2
04.11.2013 [ff978a7] fixed gzipping the response text
31.10.2013 [3e0952d] fixed definition undefined rounding constant on php 5.2
31.10.2013 [f731f59] fixed error when using func_get_args() as arguments on php 5.2
31.10.2013 [61bf630] fixed bug in compile stylesheets in js_api module when using sf_base_domain
31.10.2013 [246cf74] fixed openbasedir restriction issues
30.10.2013 [a30b053] merge debug options with defaults
30.10.2013 [8d25624] updated web debug, all panels are configurable
30.10.2013 [1ce240c] fixed prefetch task
30.10.2013 [58f8090] fixed view cache manager, return boolean
30.10.2013 [38cc432] added possiblity to compress xml output
30.10.2013 [2409ff0] removed unused parameters
30.10.2013 [ec230e9] fixed functional test, daylightsavings
30.10.2013 [d7fff60] fixed unit test - daylight savings
30.10.2013 [33ba9e5] added missing ;
30.10.2013 [12cfeb9] fixed Mail helper, when path is not absolute, look inside data_dir
30.10.2013 [b5627ee] removed unused property from sfEvent
28.10.2013 [ae02f49] fixed invalid condition when detecting relative url root
28.10.2013 [7c0a0d4] fixed #98
27.10.2013 [a387994] updated search expression to contain info about the original query is has been constructed for
26.10.2013 [4334159] updated stylings of ui autocomplete
26.10.2013 [6d10ee6] added getConfigCache() to application
26.10.2013 [347cc2b] removed search helper
26.10.2013 [9e5b7be] removed reference to search config handler
26.10.2013 [6eec08f] removed test for search source
26.10.2013 [c26b71b] fixed bug in form builder, timestamp widget is not time widget
26.10.2013 [c1a2bd8] updated copyright
26.10.2013 [d93c076] added style for ui autocomplete input
26.10.2013 [0913e36] refactored search, removed search source - will be part of plugin
25.10.2013 [7ee5208] fixed debug dump when running inside cli, added dumpVar to base cli task
24.10.2013 [72cf4d1] removed &
24.10.2013 [374cbd8] fixed #97
24.10.2013 [49c647e] added $asTranslatedString parameter to getError() to sfFormField
24.10.2013 [2dee301] fixed i18n extraction from validators
24.10.2013 [36a38ef] added removeAttribute() to user
24.10.2013 [5ba721a] fixed i18n extraction
24.10.2013 [dfb5c67] updated factories.yml to prevent merging with default section
24.10.2013 [c49aebd] added additional methods to sfIUser interface !BC
24.10.2013 [be496eb] fixed security filter
23.10.2013 [765f1d1] updated docs
23.10.2013 [ff199c0] added base app to project skeleton, updated app class inside app skeleton
23.10.2013 [efa656a] fixed plugin packaging
23.10.2013 [f545464] added possibility to enable/disable backtrace in var logger
23.10.2013 [411b5e7] added empty favicon to project skeleton
23.10.2013 [894dec1] fixed #96
23.10.2013 [7264e98] implemented storage session id verification
23.10.2013 [84e36a4] valid regenerate for pdo session storage
22.10.2013 [b6e2e72] ufff, pdo fix, going to bed
22.10.2013 [6b3a1b8] removed invalid condition
22.10.2013 [a65ac88] fixed the pdo storage with sqlite
22.10.2013 [3002534] added hash function and bits/char info to debug panel
22.10.2013 [add92e2] updated session, added support for transaction start when pdo session starts
22.10.2013 [e9d2067] fixed form debug helper, was generating invalid html
22.10.2013 [f3165ec] fixed yaml 1.2 in skeleton logging.yml
22.10.2013 [92e5cc8] disabled mail.add_x_header for security seasons
22.10.2013 [af1d2c0] added web debug, errors, exception build task
22.10.2013 [0075578] implemented sfIUserAware to the form
22.10.2013 [a9ad52c] updated logging in filter chain, service container
22.10.2013 [28412bc] refactored exceptions, production errors
22.10.2013 [8e476f0] updated web debug, added debug dumper
22.10.2013 [d791815] fixed typo in markdown parser
22.10.2013 [c831ed1] updated loggers
22.10.2013 [10b4b76] updated the debug backtrace
22.10.2013 [de8f94e] added log analyzer, fixes #58
18.10.2013 [b1d4957] added getVersion() to plugin
18.10.2013 [f6f7b99] fixed html validator
17.10.2013 [9de6e8d] added debug backtrace + decorators
17.10.2013 [0d2735b] added syntax highlighters
16.10.2013 [18d3da2] added sfLimitedScope class
15.10.2013 [13e519e] fixed #95
11.10.2013 [9c584a5] updated logging in service container, minor bug fix for undefined variable
11.10.2013 [72a96e8] updated logging in event dispatcher
08.10.2013 [c93a216] added skype username validator
08.10.2013 [80b9811] updated filters, added log() to sfFilter
08.10.2013 [edb6f8a] updated robots.txt
08.10.2013 [31322a7] updated web debug panel, panels can access event dispatcher from the web_debug class
07.10.2013 [05bdc4a] implemented sfIEventDispatcherAware to validators
07.10.2013 [fe6f4a1] implemented access to event dispatcher to the form
07.10.2013 [2a885f3] added action setting to app skeleton settings.yml
07.10.2013 [1c3915d] fixed button :focus state
07.10.2013 [40338b9] added sfDataUri + sfIDataUriConvertable, updated sfImage to use the data uri convertable intarface
04.10.2013 [bdf2f5c] fixed #93
03.10.2013 [b09fce6] fixed asset packaging test
03.10.2013 [b6c340e] fixed less compiler unit test
03.10.2013 [6f08dc7] removed compiled less + updated gitignore
03.10.2013 [12c10c9] fixed dynatree less, which causes problems with compilation with comments
03.10.2013 [40084cb] fixed #90
03.10.2013 [facd973] fixed #92
03.10.2013 [f264adf] fixed #91
02.10.2013 [d505b6f] updated the canonical_url() helper, needs refactoring of routing
02.10.2013 [afac4d9] updated sfHtml and sfWidget escaping methods

lib/helper/UrlHelper.php
<?php
/*
 * This file is part of the Sift PHP framework.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

/**
 * UrlHelper.
 *
 * @package    Sift
 * @subpackage helper
 */

/**
 * Returns a routed URL based on the module/action passed as argument
 * and the routing configuration.
 *
 * <b>Examples:</b>
 * <code>
 *  echo url_for('my_module/my_action');
 *    => /path/to/my/action
 *  echo url_for('@my_rule');
 *    => /path/to/my/action
 *  echo url_for('@my_rule', true);
 *    => http://myapp.example.com/path/to/my/action
 * </code>
 *
 * @param string $internal_uri Internal url in the format 'module/action' or '@rule' of the action
 * @param boolean $absolute Return absolute path?
 * @param array $getParameters Array of get parameters to append to the url
 * @param string $protocol
 * @return string routed URL
 */
function url_for($internal_uri, $absolute = false, $getParameters = array(), $protocol = null)
{
  static $controller;

  if(!isset($controller))
  {
    $controller = sfContext::getInstance()->getController();
  }

  return $controller->genUrl($internal_uri, $absolute, $getParameters, $protocol = null);
}

/**
 * Check if current page is homepage
 *
 * @return boolean
 */
function is_homepage()
{
  return sfRouting::getInstance()->getCurrentInternalUri(true) == '@homepage';
}

/**
 * Returns current internal uri
 *
 * @param boolea $as_safe_url Encode url to be used as safe url?
 * @param boolean $with_route_name
 * @return string
 */
function get_current_uri($as_safe_url = false, $with_route_name = true)
{
  $uri = sfRouting::getInstance()->getCurrentInternalUri($with_route_name);
  if($as_safe_url)
  {
    $uri = sfSafeUrl::encode($uri);
  }
  return $uri;
}

/**
 * Returns link to print (only for javascript)
 *
 * @param string $name Link name (default -> print)
 * @return string
 */
function link_to_print($name = 'print')
{
  $js = sprintf("document.write('<div class=\"print\"><a onclick=\"javascript:window.print();\" title=\"%s\">%s<'+'/a><'+'/div>');", __($name), __($name));
  $content = "\n//<![CDATA[\n" . $js . "\n//]]>\n";
  return '<script type="text/javascript">' . $content . '</script>';
}

/**
 * Creates a <a> link tag of the given name using a routed URL
 * based on the module/action passed as argument and the routing configuration.
 * It's also possible to pass a string instead of a module/action pair to
 * get a link tag that just points without consideration.
 * If null is passed as a name, the link itself will become the name.
 * If an object is passed as a name, the object string representation is used.
 * One of the options serves for for creating javascript confirm alerts where
 * if you pass 'confirm' => 'Are you sure?', the link will be guarded
 * with a JS popup asking that question. If the user accepts, the link is processed,
 * otherwise not.
 *
 * <b>Options:</b>
 * - 'absolute' - if set to true, the helper outputs an absolute URL
 * - 'query_string' - to append a query string (starting by ?) to the routed url
 * - 'confirm' - displays a javascript confirmation alert when the link is clicked
 * - 'popup' - if set to true, the link opens a new browser window
 * - 'post' - if set to true, the link submits a POST request instead of GET (caution: do not use inside a form)
 *
 * <b>Note:</b> The 'popup' and 'post' options are not compatible with each other.
 *
 * <b>Examples:</b>
 * <code>
 *  echo link_to('Delete this page', 'my_module/my_action');
 *    => <a href="/path/to/my/action">Delete this page</a>
 *  echo link_to('Visit Hoogle', 'http://www.hoogle.com');
 *    => <a href="http://www.hoogle.com">Visit Hoogle</a>
 *  echo link_to('Delete this page', 'my_module/my_action', array('id' => 'myid', 'confirm' => 'Are you sure?', 'absolute' => true));
 *    => <a href="http://myapp.example.com/path/to/my/action" id="myid" onclick="return confirm('Are you sure?');">Delete this page</a>
 * </code>
 *
 * @param  string name of the link, i.e. string to appear between the <a> tags
 * @param  string 'module/action' or '@rule' of the action
 * @param  array additional HTML compliant <a> tag parameters
 * @return string HTML compliant <a href> tag
 * @see    url_for
 */
function link_to($name = '', $internal_uri = '', $options = array())
{
  $html_options = _parse_attributes($options);

  $html_options = _convert_options_to_javascript($html_options);

  $absolute = false;
  if(isset($html_options['absolute_url']))
  {
    $html_options['absolute'] = $html_options['absolute_url'];
    unset($html_options['absolute_url']);
  }
  if(isset($html_options['absolute']))
  {
    $absolute = (boolean) $html_options['absolute'];
    unset($html_options['absolute']);
  }
  $protocol = null;
  if(isset($html_options['protocol']))
  {
    $protocol = (boolean) $html_options['protocol'];
    unset($html_options['protocol']);
  }
  $getParameters = array();
  if(isset($html_options['get_parameters']))
  {
    $getParameters = (array) $html_options['get_parameters'];
    unset($html_options['get_parameters']);
  }

  $html_options['href'] = url_for($internal_uri, $absolute, $getParameters, $protocol);

  if(isset($html_options['query_string']))
  {
    $html_options['href'] .= strpos($html_options['href'], '?') !== false ?
            ('&' . $html_options['query_string']) : ('?' . $html_options['query_string']);

    unset($html_options['query_string']);
  }

  if(is_object($name))
  {
    if(method_exists($name, '__toString'))
    {
      $name = $name->__toString();
    }
    else
    {
      throw new sfException(sprintf('Object of class "%s" cannot be converted to string (Please create a __toString() method)', get_class($name)));
    }
  }

  if(!strlen($name))
  {
    $name = $html_options['href'];
  }

  return content_tag('a', $name, $html_options);
}

/**
 * If the condition passed as first argument is true,
 * creates a <a> link tag of the given name using a routed URL
 * based on the module/action passed as argument and the routing configuration.
 * If the condition is false, the given name is returned between <span> tags
 *
 * <b>Options:</b>
 * - 'tag' - the HTML tag that must enclose the name if the condition is false, defaults to <span>
 * - 'absolute' - if set to true, the helper outputs an absolute URL
 * - 'query_string' - to append a query string (starting by ?) to the routed url
 * - 'confirm' - displays a javascript confirmation alert when the link is clicked
 * - 'popup' - if set to true, the link opens a new browser window
 * - 'post' - if set to true, the link submits a POST request instead of GET (caution: do not use inside a form)
 *
 * <b>Examples:</b>
 * <code>
 *  echo link_to_if($user->isAdministrator(), 'Delete this page', 'my_module/my_action');
 *    => <a href="/path/to/my/action">Delete this page</a>
 *  echo link_to_if(!$user->isAdministrator(), 'Delete this page', 'my_module/my_action');
 *    => <span>Delete this page</span>
 * </code>
 *
 * @param  bool condition
 * @param  string name of the link, i.e. string to appear between the <a> tags
 * @param  string 'module/action' or '@rule' of the action
 * @param  array additional HTML compliant <a> tag parameters
 * @return string XHTML compliant <a href> tag or name
 * @see    link_to
 */
function link_to_if($condition, $name = '', $internal_uri = '', $options = array())
{
  $html_options = _parse_attributes($options);
  if($condition)
  {
    unset($html_options['tag']);
    return link_to($name, $internal_uri, $html_options);
  }
  else
  {
    unset($html_options['query_string']);
    unset($html_options['absolute_url']);
    unset($html_options['absolute']);

    $tag = _get_option($html_options, 'tag', 'span');

    return content_tag($tag, $name, $html_options);
  }
}

/**
 * If the condition passed as first argument is false,
 * creates a <a> link tag of the given name using a routed URL
 * based on the module/action passed as argument and the routing configuration.
 * If the condition is true, the given name is returned between <span> tags
 *
 * <b>Options:</b>
 * - 'tag' - the HTML tag that must enclose the name if the condition is true, defaults to <span>
 * - 'absolute' - if set to true, the helper outputs an absolute URL
 * - 'query_string' - to append a query string (starting by ?) to the routed url
 * - 'confirm' - displays a javascript confirmation alert when the link is clicked
 * - 'popup' - if set to true, the link opens a new browser window
 * - 'post' - if set to true, the link submits a POST request instead of GET (caution: do not use inside a form)
 *
 * <b>Examples:</b>
 * <code>
 *  echo link_to_unless($user->isAdministrator(), 'Delete this page', 'my_module/my_action');
 *    => <span>Delete this page</span>
 *  echo link_to_unless(!$user->isAdministrator(), 'Delete this page', 'my_module/my_action');
 *    => <a href="/path/to/my/action">Delete this page</a>
 * </code>
 *
 * @param  bool condition
 * @param  string name of the link, i.e. string to appear between the <a> tags
 * @param  string 'module/action' or '@rule' of the action
 * @param  array additional HTML compliant <a> tag parameters
 * @return string XHTML compliant <a href> tag or name
 * @see    link_to
 */
function link_to_unless($condition, $name = '', $url = '', $options = array())
{
  return link_to_if(!$condition, $name, $url, $options);
}

/**
 * Creates an <input> button tag of the given name pointing to a routed URL
 * based on the module/action passed as argument and the routing configuration.
 * The syntax is similar to the one of link_to.
 *
 * <b>Options:</b>
 * - 'absolute' - if set to true, the helper outputs an absolute URL
 * - 'query_string' - to append a query string (starting by ?) to the routed url
 * - 'confirm' - displays a javascript confirmation alert when the button is clicked
 * - 'popup' - if set to true, the button opens a new browser window
 * - 'post' - if set to true, the button submits a POST request instead of GET (caution: do not use inside a form)
 *
 * <b>Examples:</b>
 * <code>
 *  echo button_to('Delete this page', 'my_module/my_action');
 *    => <input value="Delete this page" type="button" onclick="document.location.href='/path/to/my/action';" />
 * </code>
 *
 * @param  string name of the button
 * @param  string 'module/action' or '@rule' of the action
 * @param  array additional HTML compliant <input> tag parameters
 * @return string XHTML compliant <input> tag
 * @see    url_for, link_to
 */
function button_to($name, $internal_uri ='', $options = array())
{
  $html_options = _parse_attributes($options);
  $html_options['value'] = $name;

  if(isset($html_options['post']) && $html_options['post'])
  {
    if(isset($html_options['popup']))
    {
      throw new sfConfigurationException('You can\'t use "popup" and "post" together');
    }
    $html_options['type'] = 'submit';
    unset($html_options['post']);
    $html_options = _convert_options_to_javascript($html_options);
    return form_tag($internal_uri, array('method' => 'post', 'class' => 'button_to')) . content_tag('div', tag('button', $html_options)) . '</form>';
  }

  $url = url_for($internal_uri);
  if(isset($html_options['query_string']))
  {
    $url = $url . '?' . $html_options['query_string'];
    unset($html_options['query_string']);
  }
  $url = "'" . $url . "'";

  $html_options['type'] = 'button';

  if(isset($html_options['popup']))
  {
    $html_options = _convert_options_to_javascript($html_options, $url);
    unset($html_options['popup']);
  }
  else
  {
    $html_options['onclick'] = "document.location.href=" . $url . ";";
    $html_options = _convert_options_to_javascript($html_options);
  }

  return content_tag('button', sprintf('<span>%s</span>', $name), $html_options);
}

/**
 * Creates a <a> link tag to the given email (with href="mailto:...").
 * If null is passed as a name, the email itself will become the name.
 *
 * <b>Options:</b>
 * - 'encode' - if set to true, the email address appears with various random encoding for each letter.
 * The mail link still works when encoded, but the address doesn't appear in clear
 * in the source. Use it to prevent spam (efficiency not guaranteed).
 *
 * <b>Examples:</b>
 * <code>
 *  echo mail_to('webmaster@example.com');
 *    => <a href="mailto:webmaster@example.com">webmaster@example.com</a>
 *  echo mail_to('webmaster@example.com', 'send us an email');
 *    => <a href="mailto:webmaster@example.com">send us an email</a>
 *  echo mail_to('webmaster@example.com', 'send us an email', array('encode' => true));
 *    => <a href="
  &#x6d;a&#x69;&#x6c;&#x74;&#111;&#58;&#x77;&#x65;b&#x6d;as&#116;&#x65;&#114;
  &#64;&#101;&#x78;&#x61;&#x6d;&#x70;&#108;&#x65;&#46;&#99;&#x6f;&#109;
  ">send us an email</a>
 * </code>
 *
 * @param  string target email
 * @param  string name of the link, i.e. string to appear between the <a> tags
 * @param  array additional HTML compliant <a> tag parameters
 * @return string XHTML compliant <a href> tag
 * @see    link_to
 */
function mail_to($email, $name = '', $options = array(), $default_value = array())
{
  $html_options = _parse_attributes($options);

  $html_options = _convert_options_to_javascript($html_options);

  $default_tmp = _parse_attributes($default_value);
  $default = array();
  foreach($default_tmp as $key => $value)
  {
    $default[] = urlencode($key) . '=' . urlencode($value);
  }
  $options = count($default) ? '?' . implode('&', $default) : '';

  if(isset($html_options['encode']) && $html_options['encode'])
  {
    unset($html_options['encode']);
    $html_options['href'] = _encodeText('mailto:' . $email . $options);
    if(!$name)
    {
      $name = _encodeText($email);
    }
  }
  else
  {
    $html_options['href'] = 'mailto:' . $email . $options;
    if(!$name)
    {
      $name = $email;
    }
  }

  return content_tag('a', $name, $html_options);
}

function _convert_options_to_javascript($html_options, $url = 'this.href')
{
  // confirm
  $confirm = isset($html_options['confirm']) ? $html_options['confirm'] : '';
  unset($html_options['confirm']);

  // popup
  $popup = isset($html_options['popup']) ? $html_options['popup'] : '';
  unset($html_options['popup']);

  // post
  $post = isset($html_options['post']) ? $html_options['post'] : '';
  unset($html_options['post']);

  $onclick = isset($html_options['onclick']) ? $html_options['onclick'] : '';

  if($popup && $post)
  {
    throw new sfConfigurationException('You can\'t use "popup" and "post" in the same link');
  }
  else if($confirm && $popup)
  {
    $html_options['onclick'] = $onclick . 'if (' . _confirm_javascript_function($confirm) . ') { ' . _popup_javascript_function($popup, $url) . ' };return false;';
  }
  else if($confirm && $post)
  {
    $html_options['onclick'] = $onclick . 'if (' . _confirm_javascript_function($confirm) . ') { ' . _post_javascript_function() . ' };return false;';
  }
  else if($confirm)
  {
    if($onclick)
    {
      $html_options['onclick'] = 'if (' . _confirm_javascript_function($confirm) . ') { return ' . $onclick . '} else return false;';
    }
    else
    {
      $html_options['onclick'] = 'return ' . _confirm_javascript_function($confirm) . ';';
    }
  }
  else if($post)
  {
    $html_options['onclick'] = $onclick . _post_javascript_function() . 'return false;';
  }
  else if($popup)
  {
    $html_options['onclick'] = $onclick . _popup_javascript_function($popup, $url) . 'return false;';
  }

  return $html_options;
}

function _confirm_javascript_function($confirm)
{
  $function = 'confirm';
  $function = sfCore::getEventDispatcher()->filter(new sfEvent('core.javascript.confirm_function'),
          $function)->getReturnValue();
  return sprintf("%s('" . escape_javascript($confirm) . "')", $function);
}

function _popup_javascript_function($popup, $url = '')
{
  if(is_array($popup))
  {
    if(isset($popup[1]))
    {
      return "var w=window.open(" . $url . ",'" . $popup[0] . "','" . $popup[1] . "');w.focus();";
    }
    else
    {
      return "var w=window.open(" . $url . ",'" . $popup[0] . "');w.focus();";
    }
  }
  else
  {
    return "var w=window.open(" . $url . ");w.focus();";
  }
}

function _post_javascript_function()
{
  return "f = document.createElement('form'); document.body.appendChild(f); f.method = 'POST'; f.action = this.href; f.submit();";
}

function _encodeText($text, $type = 'hextents')
{
  if(class_exists('myMailto'))
  {
    return myMailto::encode($text, $type);
  }

  $encoded_text = '';

  for($i = 0; $i < strlen($text); $i++)
  {
    $char = $text{$i};
    $r = rand(0, 100);

    # roughly 10% raw, 45% hex, 45% dec
    # '@' *must* be encoded. I insist.
    if($r > 90 && $char != '@')
    {
      $encoded_text .= $char;
    }
    else if($r < 45)
    {
      $encoded_text .= '&#x' . dechex(ord($char)) . ';';
    }
    else
    {
      $encoded_text .= '&#' . ord($char) . ';';
    }
  }

  return $encoded_text;
}
